internal/provider/vaultsecrets/resource_vault_secrets_sync.go

```
@@ -0,0 +1,247 @@
+			WithOrganizationID(sync.OrganizationID.ValueString()).
+			WithProjectID(sync.ProjectID.ValueString()).
+			WithName(sync.Name.ValueString()), nil)
+		if err != nil && !clients.IsResponseForbidden(err) { // The HVS API returns 403 if the sync doesn't exist even if the principal has the correct permissions.
+			return nil, err
+		}
+		if response == nil || response.Payload == nil {
+			return nil, nil
+		}
+		return response.Payload.Sync, nil
+	})...)
+}
+
+func (r *resourceVaultSecretsSync) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	resp.Diagnostics.Append(decorateOperation[*Sync](ctx, r.client, &resp.State, req.Plan.Get, "creating", func(i hvsResource) (any, error) {
+		sync, ok := i.(*Sync)
+		if !ok {
+			return nil, fmt.Errorf("invalid resource type, expected *Sync, got: %T, this is a bug on the provider", i)
+		}
+
+		res, err := r.client.VaultSecrets.GetIntegration(secret_service.NewGetIntegrationParamsWithContext(ctx).
+			WithOrganizationID(sync.OrganizationID.ValueString()).
+			WithProjectID(sync.ProjectID.ValueString()).
+			WithName(sync.IntegrationName.ValueString()), nil)
+
+		if err != nil {
+			return nil, err
+		}
+
+		providerType := res.Payload.Integration.Provider
+
+		response, err := r.client.VaultSecrets.CreateSync(&secret_service.CreateSyncParams{
+			Body: &secretmodels.SecretServiceCreateSyncBody{
+				Name:             sync.Name.ValueString(),
+				IntegrationName:  sync.IntegrationName.ValueString(),
+				Type:             providerType,
+				SyncConfigGitlab: sync.gitlabConfig,
+			},
+			OrganizationID: sync.OrganizationID.ValueString(),
+			ProjectID:      sync.ProjectID.ValueString(),
+		}, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		return response.Payload.Sync, nil
+	})...)
+}
+
+func (r *resourceVaultSecretsSync) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
```

curious why the update isn't supported here?

---

ah, we don't have an update API for `Sync` https://github.com/hashicorp/cloud-vault-secrets/blob/0b11ac4fb539796a4f62d9a6f6a16457e1ca308b/internal/repository/syncs/syncs.go

---

ahh gotcha!

---

We should still handle resource config changes gracefully. If the API does not support updates all the field should cause a recreate and the update operation should throw a clear unimplemented/unsupported error if it is triggered by accident.

For example if I change my Gitlab scope from project to group I get this error:
```
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to hcp_vault_secrets_sync.my_gitlab_sync, provider "provider[\"registry.terraform.io/hashicorp/hcp\"]" produced an unexpected new value: .gitlab_config.scope: was
│ cty.StringVal("GROUP"), but now cty.StringVal("PROJECT").
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```